# **chapter6. 예외 처리**

## **06-1 구문 오류와 예외**

## **오류의 종류**

    - 구문오류: 실행 전에 발생하는 오류
    - 예외(런타임 오류): 프로그램 실행 중에 발생하는 오류

-  **구문오류**

   괄호의 개수, 들여쓰기 문제 등으로 인해 프로그램이 실행되기도 전에 발생하는 오류

* * *

구문오류가 발생하는 코드

```py
# 프로그램 시작
print("# 프로그램이 시작되었습니다!")
# 구문 오류 발생코드
print("# 예외를 강제로 발생시켜 볼게요!)
```

SyntaxError: 구문에 에러가 있어 프로그램이 실행조차 되지 않는 오류

- **예외**

  프로그램이 실행된 다음, 실행 중에 발생하는 오류

## **기본 예외 처리**

**예외처리**: 예외를 해결하는 모든 것

#### **예외를 처리하는 방법**

1. 조건문을 사용하는 방법(기본 예외 처리)
2. try 구문을 사용하는 방법 

## **try except 구문**

- try except 구문: 프로그래밍 언어에서 예외를 처리할 수 있는 구문

```
try: 
    예외가 발생할 가능성이 있는 코드
except: 
    예외가 발생했을 때 실행할 코드
```

## **try except else 구문**

try except 구문 뒤에 else 구문을 붙여서 사용하면 '예외가 발생하지 않았을 때 실행할 코드'를 지정할 수 있다. 

```
try:
    예외가 발생할 가능성이 있는 코드
except:
    예외가 발생했을 때 실행할 코드
else:
    예외가 발생하지 않았을 때 실행할 코드
```

## **finally 구문**

finally 구문은 예외 처리 구문에서 가장 마지막에 사용할 수 있는 구문

```
try:
    예외가 발생할 가능성이 있는 코드
except:
    예외가 발생했을 때 실행할 코드
else:
    예외가 발생하지 않았을 때 실행할 코드
finally:
    무조건 실행할 코드
```

### **try, except, finally 구문의 조합**
- try 구문은 단독으로 사용할 수 없으며, 반드시 except 구문 또는 finally 구문과 함께 사용해야 한다. 
- else 구문은 반드시 except 구문 뒤에 사용해야 한다.

* * *

## **06-2 예외 고급**

**예외 객체**: 예외와 관련된 정보를 담고 있는 객체

### **예외 구분하기**

예외 객체를 사용하면 except 구문을 if 조건문처럼 사용해서 예외를 구분할 수 있다. 

```
try:
    예외가 발생할 가능성이 있는 구문
except 예외의 종류A:
    예외A가 발생했을 때 실행할 구문
except 예외의 종류B:
    예외B가 발생했을 때 실행할 구문
except 예외의 종류C:
    예외C가 발생했을 때 실행할 구문
```

### **raise 구문**

#### 아직 구현되지 않은 부분에서 강제로 예외 발생시키기

```py
# 입력을 받습니다
number = input("정수 입력> ")
number = int(number)
# 조건문 사용
if number > 0:
    # 양수일 때: 아직 미구현 상태입니다. 
    raise NotImp ementedError
else:
    # 음수일 때: 아직 미구현 상태입니다. 
    raise NotImp ementedErrorr
```
아직 구현되지 않은 부분이므로 일부러 예외를 발생시켜 프로그램을 죽게 만들어 잊지 않도록 하는 것.

**raise 키워드**: 예외를 강제로 발생시키는 기능
